---
title: "Downstream Analysis Following Applying vmrseq"
author:
- name: Ning Shen
  affiliation: 
  - Department of Statistics, University of British Columbia; 
  - Centre for Molecular Medicine and Therapeutics, BC Children's Hospital Research Institute
  email: ning.shen.wk@gmail.com
output: 
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    number_sections: true
    fig_width: 9
    fig_height: 4.5
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "##"
)
```


# Background

In this vignette, we assume that scripts of the previous steps has been run, i.e.,

- `01_run_vmrseq.R`: Apply vmrseq to the input dataset. The most important functions called in this script is `vmrseqSmooth` and `vmrseqFit`, which are used to detect the VMRs from the input dataset.
- `02_run_region_summary.R`: Summarize regional information. The most important function called in this script is `regionSummary`, which provides total covered cells and methylated cells per regions, as well as regional average methylation.

Feel free to check out the scripts! A brief instroduction of the input dataset can be found in the README of this repository.

# Import the results from `vmrseq`

We import the summarized regional information that are output by `02_run_region_summary.R`:
```{r}
regions.se <- HDF5Array::loadHDF5SummarizedExperiment(here::here("data", "02_run_region_summary", "vmrs_summary"))
```

We store the genomic coordinates in the object `regions.gr` and the regional average methylation matrix into object `MF` for further use.
```{r}
regions.gr <- GenomicRanges::granges(regions.se)
MF <- SummarizedExperiment::assays(regions.se)$MF |> as.matrix()
```

# Gene / CpG context annotation

Genomic regions resulting from next-generation sequencing experiments and bioinformatics pipelines are more meaningful when annotated to genomic features. Using the `annotatr` package, we are able to examine which genomic features our regions of interest (i.e., VMRs in this example) are enriched in. 

## Util functions

These are the wrapper functions that will be used to generate annotation figures:
```{r}
getRandomizedAnnotation <- function(regions, annotations) {
  
  # Generate random regions with the same widths and chromosomes as our regions of interest
  set.seed(100)
  rand_regions <- annotatr::randomize_regions(
    regions = regions,
    allow.overlaps = TRUE,
    per.chromosome = TRUE)
  
  # Annotate the random regions using the same annotations as above
  rand_annt <- annotatr::annotate_regions(
    regions = rand_regions,
    annotations = annotations,
    ignore.strand = TRUE,
    quiet = TRUE)
  
  return(rand_annt)
}

plotAnnotations <- function(regions, annotations, plot_title) {
  
  ## Obtain annotations for the regions of interest 
  region_annt <- annotatr::annotate_regions(
    regions = regions,
    annotations = annotations,
    ignore.strand = TRUE,
    quiet = FALSE
  ) |> data.frame()
  
  ## Obtain annotations for randomized regions 
  random_annt <- getRandomizedAnnotation(regions = regions, annotations = annotations)
  
  # Plot the comparison of gene context between our regions of interest and random regions
  annotatr::plot_annotation(
    annotated_regions = region_annt,
    annotated_random = random_annt,
    annotation_order = unique(annotations$type),
    plot_title = plot_title,
    x_label = 'Annotations',
    y_label = 'Count')
}
```

Specify that our targeted genome assembly is "mm10":
```{r}
GenomeInfoDb::genome(regions.gr) <- "mm10"
```


## Gene annotation

Before annotating regions, they must be built with `build_annotations()` with desired annotation codes.
```{r message=FALSE}
types_gene <- grep("mm10_genes.*", annotatr::builtin_annotations(), value = T)
annt_genes <- annotatr::build_annotations(genome = 'mm10', annotations = types_gene)
```

We plot the distribution of gene contexts that overlap with our VMRs (bars in black) against randomized regions with the same width as our VMRs (bars in grey). It may be observed that the VMRs selected by vmrseq slightly favors towards genic regions over intergenic ones.
```{r}
plotAnnotations(regions = regions.gr, annotations = annt_genes, plot_title = "Gene Context Annotation of VMRs")
```

## CpG annotation

Similar to gene context, we also plot the distribution of CpG contexts that overlap with our VMRs (bars in black) against randomized regions with the same width as our VMRs (bars in grey). 

```{r}
types_cpg <- "mm10_cpgs"
annt_cpgs <- annotatr::build_annotations(genome = 'mm10', annotations = types_cpg)
```

```{r}
plotAnnotations(regions = regions.gr, annotations = annt_cpgs, plot_title = "CpG Context Annotation of VMRs")
```


# Unsupervised cell clustering analysis

In this section we demonstrate the procedure of an unsupervised cell clustering analysis using the detected VMRs, as similarly presented in the vmrseq paper. 

## Compute cell-to-cell dissimilarity distance matrix

We first compute a cell-to-cell dissimilarity matrix based on the `MF` assay output by `region.summary`. The `daisy` function from `cluster` is used here for it allows NA values in its input.

```{r}
d_mat <- cluster::daisy(t(MF), metric = 'manhattan', stand = FALSE, warnBin = FALSE) |> as.matrix()
```


## Cluster with UMAP

Then, UMAP coordinates of the cells were computed using the `umap` function from package `uwot`:
```{r}
umap.df <- uwot::umap(d_mat |> as("sparseMatrix"), n_neighbors = 15, n_components = 2) 
colnames(umap.df) <- c('UMAP 1', 'UMAP 2')
```

We plot the cells with the UMAP coordinates in the following figure. We may observe that the two broad classes are a primary source of heterogeneity in this example dataset, as they are far apart in terms of UMAP coords. The subtypes within those broad classes are also fairly separated, demonstrating that the VMRs detected by vmrseq has captures heterogeneity associated with the subtypes as well.
```{r}
library(ggplot2)
set.seed(2024)
plot.df <- cbind(umap.df, as.data.frame(SummarizedExperiment::colData(regions.se)))
plot.df |>
  ggplot(aes(`UMAP 1`, `UMAP 2`, shape = broad_class, color = subtype)) +
  geom_point() +
  scale_shape_manual(values = c(1, 2)) + 
  theme_classic() 
```


